plugin/nomad/nomad.go

```diff
@@ -0,0 +1,149 @@
+package nomad
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/metrics"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+	"github.com/hashicorp/nomad/api"
+	nomad "github.com/hashicorp/nomad/api"
+
+	"github.com/miekg/dns"
+)
+
+const pluginName = "nomad"
+
+var (
+	log        = clog.NewWithPlugin(pluginName)
+	defaultTTL = time.Duration(30 * time.Second).Seconds()
+)
+
+// Nomad is a plugin that serves records for Nomad services
+type Nomad struct {
+	Next plugin.Handler
+
+	ttl    uint32
+	client *nomad.Client
+}
+
+// ServeDNS implements the plugin.Handler interface. This method gets called when example is used
+// in a Server.
+func (n Nomad) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	qname := state.Name()
+	qtype := state.QType()
+
+	// Split the query name with a `.` as the delimiter and extract namespace and service name.
+	// If the query is not for a Nomad service, return.
+	qnameSplit := dns.SplitDomainName(qname)
+	if len(qnameSplit) < 3 || qnameSplit[2] != "nomad" {
+		return plugin.NextOrFailure(n.Name(), n.Next, ctx, w, r)
+	}
+	namespace := qnameSplit[1]
+	serviceName := qnameSplit[0]
+
+	m := new(dns.Msg)
+	m.SetReply(r)
+	m.Authoritative = true
+	m.Compress = true
+	m.Rcode = dns.RcodeSuccess
+	m.Answer = []dns.RR{}
+	m.Extra = []dns.RR{}
+
+	header := dns.RR_Header{
+		Name:   state.QName(),
+		Rrtype: state.QType(),
+		Class:  dns.ClassINET,
+		Ttl:    n.ttl,
+	}
+
+	// Fetch service registrations for the given service.
+	log.Debugf("Looking up record for svc: %s namespace: %s", serviceName, namespace)
+	svcRegistrations, _, err := n.client.Services().Get(serviceName, (&api.QueryOptions{Namespace: namespace}))
+	if err != nil {
+		m.Rcode = dns.RcodeServerFailure
+		w.WriteMsg(m)
+		requestFailedCount.WithLabelValues(metrics.WithServer(ctx), namespace).Inc()
+		return dns.RcodeServerFailure, fmt.Errorf("error fetching service detail: %w", err)
```

Comment:
See my other comment about writing two response to the client. dns.RcodeServerFailure is also categorized as a failure.